wallet-api/src/Wallet/Emulator/Http.hs

```diff
@@ -0,0 +1,164 @@
+  = WalletAPIError WAPI.WalletAPIError
+  | HttpError ServantErr
+
+wallets ::
+     (MonadError ServantErr m, MonadReader State m, MonadIO m) => m [Wallet]
+wallets = do
+  var <- asks getWallets
+  ws <- liftIO $ readTVarIO var
+  pure $ Wallet <$> HM.keys ws
+
+fetchWallet ::
+     (MonadError ServantErr m, MonadReader State m, MonadIO m)
+  => Int
+  -> m Wallet
+fetchWallet walletId = do
+  var <- asks getWallets
+  ws <- liftIO $ readTVarIO var
+  if HM.member walletId ws
+    then pure (Wallet walletId)
+    else throwError err404
+
+createWallet :: (MonadReader State m, MonadIO m) => Wallet -> m NoContent
+createWallet wallet = do
+  var <- asks getWallets
+  let walletState = emptyWalletState wallet
+  liftIO . atomically $
+    modifyTVar var (HM.insert (getWallet wallet) walletState)
+  pure NoContent
+
+createPaymentWithChange ::
+     (MonadReader State m, MonadIO m, MonadError ServantErr m)
+  => Int
+  -> Value
+  -> m (Set.Set TxIn', TxOut')
+createPaymentWithChange = runWalletApiAction WAPI.createPaymentWithChange
+
+payToPublicKey ::
+     (MonadReader State m, MonadIO m, MonadError ServantErr m)
+  => Int
+  -> Value
+  -> m TxOut'
+payToPublicKey = runWalletApiAction WAPI.payToPublicKey
+
+submitTxn ::
+     (MonadReader State m, MonadIO m, MonadError ServantErr m)
+  => Int
+  -> Tx
+  -> m ()
+submitTxn = runWalletApiAction WAPI.submitTxn
+
+runWalletApiAction ::
+     (MonadReader State m, MonadIO m, MonadError ServantErr m)
+  => (a -> EmulatedWalletApi b)
+  -> Int
+  -> a
+  -> m b
+runWalletApiAction action walletId a = do
+  var <- asks getWallets
+  finalResult <-
+    liftIO . atomically $ do
+      states <- readTVar var
+      let mState = HM.lookup walletId states
+      case mState of
+        Nothing -> pure . Left . HttpError $ err404
+        Just oldState -> do
+          let (res, _) =
+                runWriter .
+                flip runStateT oldState . runExceptT . runEmulatedWalletApi $
+                action a
+          case res of
+            (Left e, _) -> pure . Left . WalletAPIError $ e
+            (Right v, newState) -> do
+              modifyTVar var (HM.insert walletId newState)
+              pure $ Right v
+  case finalResult of
+    Left (HttpError e) -> throwError e
+    Left (WalletAPIError e) ->
+      throwError $ err500 {errBody = BSL.pack . show $ e}
+    Right res -> pure res
+
+-- TODO: This will be part of the transactions API
+getTransactions ::
+     (MonadReader State m, MonadIO m, MonadError ServantErr m) => m [Tx]
+getTransactions = undefined
+
+-- | Concrete monad stack for server server
+newtype M a = M
+  { unM :: ReaderT State (ExceptT ServantErr IO) a
+  } deriving ( Functor
+             , Applicative
+             , Monad
+             , MonadReader State
+             , MonadIO
+             , MonadError ServantErr
+             )
+
+runM :: State -> M ~> Handler
+runM state r = do
+  res <- liftIO . runExceptT . flip runReaderT state . unM $ r
+  case res of
```

Comment:
`liftEither`

Reply:
dah! that's what I was looking for, can never remember it